cli/core/test/e2e/framework/plugin_operations.go

```diff
@@ -0,0 +1,267 @@
+		if err != nil {
+			errs[i] = err
+			continue
+		}
+		pm.pluginBinaryFilePath = pluginBinaryFilePath
+		pluginBinaryFilePaths[i] = pm.pluginBinaryFilePath
+	}
+	return pluginBinaryFilePaths, errs
+}
+
+// PublishPluginBinary publishes the plugin binaries to given registry bucket and returns the plugin distribution urls
+func (po *localOCIPluginOps) PublishPluginBinary(pluginsMeta []*PluginMeta) (discoveryUrls []string, errs []error) {
+	size := len(pluginsMeta)
+	distributionUrls := make([]string, size)
+	errs = make([]error, size)
+	for i, pm := range pluginsMeta {
+		// Set registry discovery url if not set already
+		if pm.registryDiscoveryURL == "" {
+			pm.registryDiscoveryURL = filepath.Join(po.registry.GetRegistryURLWithDefaultCLIPluginsBucket(), ("/" + pm.name + "/"))
+		}
+		imageRegistryURL, err := po.imgpkgOps.PushBinary(pm.pluginBinaryFilePath, pm.registryDiscoveryURL)
+		if err != nil {
+			errs[i] = err
+			continue
+		}
+		pm.binaryDistributionURL = imageRegistryURL
+		distributionUrls[i] = imageRegistryURL
+	}
+	return distributionUrls, errs
+}
+
+// GeneratePluginBundle generates plugin bundle in local file system for given plugin metadata
+func (po *localOCIPluginOps) GeneratePluginBundle(pluginsMeta []*PluginMeta) ([]string, []error) {
+	pluginsProcessed := make(map[string]bool)
+	size := len(pluginsMeta)
+	pluginBundlePath := make([]string, size)
+	errs := make([]error, size)
+	for i, pm := range pluginsMeta {
+		if _, exists := pluginsProcessed[pm.name]; exists {
+			errs[i] = errors.New("plugin name already exists, currently multiple versions of same plugin not supported")
+			continue
+		}
+		pluginsProcessed[pm.name] = true
+
+		// Set registry discovery url if not set already
+		if pm.registryDiscoveryURL == "" {
+			pm.registryDiscoveryURL = filepath.Join(po.registry.GetRegistryURLWithDefaultCLIPluginsBucket(), ("/" + pm.name + "/"))
+		}
+
+		if pm.binaryDistributionURL == "" {
+			errs[i] = errors.New("plugin binary distribution url is empty")
+			continue
+		}
+
+		pluginOverlayObj, err := po.generatePluginDiscoveryOverlay(pm)
+		if err != nil {
+			errs[i] = err
+			continue
+		}
+
+		pluginBundlePathLocal, err := po.createLocalPluginBundle(pm, pluginOverlayObj)
+		if err != nil {
+			errs[i] = err
+			continue
+		}
+
+		pluginBundlePath[i] = pluginBundlePathLocal
+	}
+	return pluginBundlePath, errs
+}
+
+// PublishPluginBundle publishes the plugin bundles to given registry bucket and returns the plugins discovery urls
+func (po *localOCIPluginOps) PublishPluginBundle(pluginsMeta []*PluginMeta) ([]string, []error) {
+	size := len(pluginsMeta)
+	discoveryUrls := make([]string, size)
+	errs := make([]error, size)
+	for i, pm := range pluginsMeta {
+		// Set registry discovery url if not set already
+		if pm.registryDiscoveryURL == "" {
+			pm.registryDiscoveryURL = filepath.Join(po.registry.GetRegistryURLWithDefaultCLIPluginsBucket(), ("/" + pm.name + "/"))
```

Comment:
consider plugin-name/target as a uniqueness to identify plugins.

Reply:
Done